Add support of pr explicit testing via /test

chmouel opened the pull request:
# Changes <!-- 🎉🎉🎉 Thank you for the PR!!! 🎉🎉🎉 -->

Users now have the ability to explicitly start a PipelineRun using the /test <pipelinerun-name> comment, irrespective of whether the pipelinerun matches the annotations. This feature grants users explicit control over the execution of PipelineRuns for specific types of Pull Requests, allowing manual initiation instead of relying solely on events.

This enhancement is particularly useful for scenarios where users need fine-grained control over the testing process, such as for Pull Requests managed by users rather than events.

Add e2e for gitlab/github(ghe)/gitea, but supports bitbutcketcloud as well

See: https://issues.redhat.com/browse/SRVKP-3561

## Demo
https://github.com/openshift-pipelines/pipelines-as-code/assets/98980/54762174-da6b-4323-ba49-47a3c4b93be0



<!-- Describe your changes here- ideally you can get that description straight from
your descriptive commit message(s)! -->

# Submitter Checklist

- [ ] 📝 Please ensure your commit message is clear and informative. For guidance on crafting effective commit messages, refer to the How to write a git commit message guide. We prefer the commit message to be included in the PR body itself rather than a link to an external website (ie: Jira ticket).

- [ ] ♽ Before submitting a PR, run make test lint to avoid unnecessary CI processing. For an even more efficient workflow, consider installing [pre-commit](https://pre-commit.com/) and running pre-commit install in the root of this repository.

- [ ] ✨ We use linters to maintain clean and consistent code. Please ensure you've run make lint before submitting a PR. Some linters offer a --fix mode, which can be executed with the command make fix-linters (ensure [markdownlint](https://github.com/DavidAnson/markdownlint) and [golangci-lint](https://github.com/golangci/golangci-lint) tools are installed first).

- [ ] 📖 If you're introducing a user-facing feature or changing existing behavior, please ensure it's properly documented.

- [ ] 🧪 While 100% coverage isn't a requirement, we encourage unit tests for any code changes where possible.

- [ ] 🎁 If feasible, please check if an end-to-end test can be added. See [README](https://github.com/openshift-pipelines/pipelines-as-code/blob/main/test/README.md) for more details.

- [ ] 🔎 If there's any flakiness in the CI tests, don't necessarily ignore it. It's better to address the issue before merging, or provide a valid reason to bypass it if fixing isn't possible (e.g., token rate limitations).

chmouel commented:
/retest

chmouel commented:
there is an issue with the dogfood cluster 
![image](https://github.com/openshift-pipelines/pipelines-as-code/assets/98980/eafdc418-8dbb-4b7b-9129-9bbd9fee3bbf)

chmouel commented:
/test go-testing

chmouel commented:
There is more work to be done

chmouel commented:
Some tests who weren't failing before (can't figure out why) are failing now, so had to fix them in commit https://github.com/openshift-pipelines/pipelines-as-code/pull/1560/commits/944143e67629fde595c7cb2b3b3211abbe92a5fe

chmouel commented:
woo woo green e2e 😅

savitaashture commented on docs/content/docs/guide/running.md:
````diff
@@ -141,6 +141,11 @@ roses are red, violets are blue. pipeline are bound to flake by design.
 /test <pipelinerun-name>
 ```
 
+Note that the `/test <pipelinerun-name>` comment will start the `PipelineRun`
````
Can we have a subsection specifying this part of the doc is to trigger not based on events something like that 

I mean

```
### GitOps command on pull or merge request

#### Event-Based Triggers
....
(Describe GitOps commands triggered by specific events, such as pull requests or merges.)

#### Non-Event-Based Triggers
...
(Specify GitOps commands that are not triggered by specific events, providing details on how they can be initiated, e.g., through `/test` comments.)
```

savitaashture commented on docs/content/docs/guide/running.md:
```diff
@@ -141,6 +141,11 @@ roses are red, violets are blue. pipeline are bound to flake by design.
+regardless if the pipelinerun is matching the annotations. This let you have
+explicit control of PipelineRun for some type of Pull Requests controlled by the
+user instead of events.
+
```
I tried to test manually 

couple of things

1. I had pipelinerun in `.tekton` dir which have 
```
metadata:
  name: another-pr-pac-demo
annotations:
    pipelinesascode.tekton.dev/on-event: "[pull_request, push]"
    pipelinesascode.tekton.dev/on-target-branch: "[main1]"
```
but gave wrong values so CI did not trigger then as per the doc added comment
```
/test another-pr-pac-demo
```
But CI did not start

Then commented those annotation and did `/test another-pr-pac-demo` then CI started

So can we explicitly add that details saying no annotations should be present in pipelinerun :thinking: or let me know if my understanding is wrong here

2. I see `/retest` is not working if its not supported can we add that info as well ?

chmouel replied:
I see yeah, thanks.. that would be a bug i need to fix i think  instead of just documentation..

chmouel commented:
/test

chmouel committed:
Add support to run PR without an event matching

Users now have the ability to explicitly start a PipelineRun using the
/test comment, irrespective of whether the pipelinerun matches the
annotations. This feature grants users explicit control over the
execution of PipelineRuns for specific types of Pull Requests, allowing
manual initiation instead of relying solely on events.

This enhancement is particularly useful for scenarios where users need
fine-grained control over the testing process, such as for Pull Requests
managed by users rather than events.

Add e2e for gitlab/github(ghe)/gitea

Jira: https://issues.redhat.com/browse/SRVKP-3561

Signed-off-by: Chmouel Boudjnah <chmouel@redhat.com>

chmouel commented:
reworked the whole pr based on-comment, kept the e2e

savitaashture commented:
PR looks good to me 
Tested locally and it works fine

@chmouel i just have one open comment https://github.com/openshift-pipelines/pipelines-as-code/pull/1560#discussion_r1462872102

chmouel commented:
thanks for the review..

i am not sure about that docs change, but maybe i'll add a section then.

savitaashture commented:
> thanks for the review..
> 
> i am not sure about that docs change, but maybe i'll add a section then.

Yup that works for me

chmouel commented:
I reworded the documentations @savitaashture please have a look

savitaashture commented on docs/content/docs/guide/running.md:
```diff
@@ -114,14 +114,14 @@ entire suite of checks once again.
-restarted :
+For example, you want to restart all your pipelineruns you can add a comment
+starting with `/retest` and all PipelineRuns attached to that pull or merge
+request will be restarted :
```
Looks good to me
just one more nit

How about including one more section under GitOps commands for pull or merge request
something like below

## GitOps commands

The GitOps commands are a way to trigger Pipelines-as-Code actions via comments
on a `Pull Request` or comment on a `Push`ed commit.

### GitOps commands on pull or merge request

If you want to trigger a GitOps command on a pull or merge request, you can include the
`GitOps` comments inside your pull or merge request. These comments can be used to
restart either all pipelines or specific ones.

For example, you want to restart all your pipelineruns you can add a comment
starting with `/retest` and all PipelineRuns attached to that pull or merge
request will be restarted :

chmouel commented:
@savitaashture updated please take a look

savitaashture reviewed (approved):
LGTM :+1: @piyush-garg do you want to take a final look

piyush-garg commented:
/lgtm
